File: LeetCode/Python Solutions/Validate Binary Tree Nodes/validate_binary_tree_nodes.py
from collections import deque

class Solution:
    def validateBinaryTreeNodes(self, n, leftChild, rightChild):
        # Step 1: Find the root (the node that is not a child of any other node)
        non_root_nodes = set()
        for i in range(n):
            if leftChild[i] != -1:
                non_root_nodes.add(leftChild[i])
            if rightChild[i] != -1:
                non_root_nodes.add(rightChild[i])

        # There should be exactly one node that is not in the non_root_nodes set
        if len(non_root_nodes) != n - 1:
            return False

        # Identify the root. It's the node not present in non_root_nodes.
        root = -1
        for i in range(n):
            if i not in non_root_nodes:
                root = i
                break

        # Step 2: Perform a BFS traversal to check for cycles and single parent property.
        visited = set()
        queue = deque([root])

        while queue:
            node = queue.popleft()
            if node in visited:
                # Cycle detected
                return False
            visited.add(node)

            # Add children to the queue
            if leftChild[node] != -1:
                queue.append(leftChild[node])
            if rightChild[node] != -1:
                queue.append(rightChild[node])

        # If all nodes are visited, it's a valid tree
        return len(visited) == n

File: LeetCode/Java Solutions/Validate Binary Tree Nodes/validate_binary_tree_nodes.java
import java.util.*;

class Solution {
    public boolean validateBinaryTreeNodes(int n, int[] leftChild, int[] rightChild) {
        // Step 1: Find the root (node not a child of any other node)
        Set<Integer> nonRootNodes = new HashSet<>();
        for (int i = 0; i < n; ++i) {
            if (leftChild[i] != -1) nonRootNodes.add(leftChild[i]);
            if (rightChild[i] != -1) nonRootNodes.add(rightChild[i]);
        }
        
        // There should be exactly one node that is not in nonRootNodes set
        if (nonRootNodes.size() != n - 1) return false;
        
        // Identify the root. It's the node not present in nonRootNodes.
        int root = -1;
        for (int i = 0; i < n; ++i) {
            if (!nonRootNodes.contains(i)) {
                root = i;
                break;
            }
        }

        // Step 2: Perform a BFS traversal to check for cycles and single parent property.
        Set<Integer> visited = new HashSet<>();
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            int node = queue.poll();
            if (visited.contains(node)) {
                // Cycle detected
                return false;
            }
            visited.add(node);

            // Add children to the queue
            if (leftChild[node] != -1) queue.offer(leftChild[node]);
            if (rightChild[node] != -1) queue.offer(rightChild[node]);
        }

        // If all nodes are visited, it's a valid tree
        return visited.size() == n;
    }
}
